cmd/jujud/bootstrap_test.go

```diff
@@ -140,6 +150,67 @@ func (s *BootstrapSuite) writeDownloadedTools(c *gc.C, tools *tools.Tools) {
 	c.Assert(err, jc.ErrorIsNil)
 }
 
+func (s *BootstrapSuite) writeDownloadedGUI(c *gc.C, gui *tools.GUIArchive) {
+	guiDir := filepath.FromSlash(agenttools.SharedGUIDir(s.dataDir))
+	err := os.MkdirAll(guiDir, 0755)
+	c.Assert(err, jc.ErrorIsNil)
+	data, err := json.Marshal(gui)
+	c.Assert(err, jc.ErrorIsNil)
+	err = ioutil.WriteFile(filepath.Join(guiDir, "downloaded-gui.txt"), data, 0644)
+	c.Assert(err, jc.ErrorIsNil)
+}
+
+func (s *BootstrapSuite) TestGUIArchiveInfoError(c *gc.C) {
+	dir := filepath.FromSlash(agenttools.SharedGUIDir(s.dataDir))
+	info := filepath.Join(dir, "downloaded-gui.txt")
+	err := os.Remove(info)
+	c.Assert(err, jc.ErrorIsNil)
+	_, cmd, err := s.initBootstrapCommand(c, nil, "--model-config", s.b64yamlEnvcfg, "--instance-id", string(s.instanceId))
+	c.Assert(err, jc.ErrorIsNil)
+	// TODO frankban: this must return an error before the feature branch is
+	// merged into master.
+	err = cmd.Run(nil)
+	c.Assert(err, jc.ErrorIsNil)
+}
+
+func (s *BootstrapSuite) TestGUIArchiveError(c *gc.C) {
+	dir := filepath.FromSlash(agenttools.SharedGUIDir(s.dataDir))
+	archive := filepath.Join(dir, "gui.tar.bz2")
+	err := os.Remove(archive)
+	c.Assert(err, jc.ErrorIsNil)
+	_, cmd, err := s.initBootstrapCommand(c, nil, "--model-config", s.b64yamlEnvcfg, "--instance-id", string(s.instanceId))
+	c.Assert(err, jc.ErrorIsNil)
+	err = cmd.Run(nil)
+	c.Assert(err, gc.ErrorMatches, "cannot read GUI archive: .*")
+}
+
+func (s *BootstrapSuite) TestGUIArchiveSuccess(c *gc.C) {
+	_, cmd, err := s.initBootstrapCommand(c, nil, "--model-config", s.b64yamlEnvcfg, "--instance-id", string(s.instanceId))
+	c.Assert(err, jc.ErrorIsNil)
+	err = cmd.Run(nil)
+	c.Assert(err, jc.ErrorIsNil)
+
+	// Retrieve the state so that it is possible to access the GUI storage.
+	st, err := state.Open(testing.ModelTag, &mongo.MongoInfo{
+		Info: mongo.Info{
+			Addrs:  []string{gitjujutesting.MgoServer.Addr()},
+			CACert: testing.CACert,
+		},
+		Password: testPasswordHash(),
+	}, mongo.DefaultDialOpts(), environs.NewStatePolicy())
+	c.Assert(err, jc.ErrorIsNil)
+	defer st.Close()
+
+	// The GUI archive has been uploaded to the GUI storage.
+	storage, err := st.GUIStorage()
+	c.Assert(err, jc.ErrorIsNil)
+	defer storage.Close()
+	allMeta, err := storage.AllMetadata()
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(allMeta, gc.HasLen, 1)
+	c.Assert(allMeta[0].Version, gc.Equals, "2.0.42")
+}
```

Comment:
AFAICT that's not jujud bootstrap responsibility, and it's tested elsewhere (cloudconfig/userdatacfg_test.go).